pkg/frontend/querymiddleware/remote_read.go

```diff
@@ -229,14 +291,46 @@ func (r *remoteReadQueryRequest) WithQuery(_ string) (MetricsQueryRequest, error
 	panic("not implemented")
 }
 
-func (r *remoteReadQueryRequest) WithHeaders([]*PrometheusHeader) MetricsQueryRequest {
+func (r *remoteReadQueryRequest) WithHeaders(_ []*PrometheusHeader) MetricsQueryRequest {
 	panic("not implemented")
 }
 
-func (r *remoteReadQueryRequest) WithStartEnd(_ int64, _ int64) (MetricsQueryRequest, error) {
-	panic("not implemented")
+// WithStartEnd clones the current remoteReadQueryRequest with a new start and end timestamp.
+func (r *remoteReadQueryRequest) WithStartEnd(start int64, end int64) (MetricsQueryRequest, error) {
+	clonedQuery, err := cloneRemoteReadQuery(r.query)
+	if err != nil {
+		return nil, err
+	}
+
+	clonedQuery.StartTimestampMs = start
+	clonedQuery.EndTimestampMs = end
+
+	if clonedQuery.Hints != nil && clonedQuery.Hints.StartMs != 0 {
```

Comment:
```suggestion
	if clonedQuery.Hints != nil && clonedQuery.Hints.StartMs < start {
```

Reply:
Done in https://github.com/grafana/mimir/pull/8374/commits/7f2ab735e566e2d1ff2488231175cf355ef497fd